File: backend/app/models/portfolio.py
"""Portfolio and Holding ORM models."""
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Numeric, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Portfolio(Base):
    __tablename__ = "portfolios"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str] = mapped_column(String(500), default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    user: Mapped["User"] = relationship("User", back_populates="portfolios")
    holdings: Mapped[list["Holding"]] = relationship(
        "Holding", back_populates="portfolio", cascade="all, delete-orphan"
    )


class Holding(Base):
    __tablename__ = "holdings"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    portfolio_id: Mapped[int] = mapped_column(ForeignKey("portfolios.id"), nullable=False, index=True)
    ticker: Mapped[str] = mapped_column(String(10), nullable=False, index=True)
    shares: Mapped[Decimal] = mapped_column(Numeric(precision=18, scale=6))
    avg_cost: Mapped[Decimal] = mapped_column(Numeric(precision=18, scale=4))
    purchased_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    portfolio: Mapped["Portfolio"] = relationship("Portfolio", back_populates="holdings")

File: frontend/src/api/portfolio.ts
import { apiClient } from './client';

export const fetchPortfolios = async () => {
  const { data } = await apiClient.get('/portfolio/');
  return data;
};

export const createPortfolio = async (payload: { name: string; description?: string }) => {
  const { data } = await apiClient.post('/portfolio/', payload);
  return data;
};

export const fetchPortfolio = async (id: number) => {
  const { data } = await apiClient.get(`/portfolio/${id}`);
  return data;
};

export const deletePortfolio = async (id: number) => {
  const { data } = await apiClient.delete(`/portfolio/${id}`);
  return data;
};

export const addHolding = async (
  portfolioId: number,
  payload: { ticker: string; shares: number; avg_cost: number }
) => {
  const { data } = await apiClient.post(`/portfolio/${portfolioId}/holdings`, payload);
  return data;
};
